horovod/spark/lightning/remote.py

```diff
@@ -74,46 +75,65 @@ def RemoteTrainer(estimator, metadata, ckpt_bytes, run_id, dataset_idx, train_ro
 
-    # disable call back for now. Because petastorm can not reset index during training.
-    callbacks = None #_make_callbacks()
+    def train(serialized_model, num_gpus):
+        import horovod.torch as hvd
+        # Horovod: initialize library.
+        # hvd.init()
 
-    def train(serialized_model):
         with tempfile.TemporaryDirectory() as last_ckpt_dir, remote_store.get_local_output_dir() as run_output_dir:
             last_ckpt_file = os.path.join(last_ckpt_dir, 'last.ckpt')
             if ckpt_bytes:
                 with open(last_ckpt_file, 'wb') as f:
                     f.write(ckpt_bytes)
 
+            # TODO: Pass the logger from estimator constructor
             logs_path = os.path.join(run_output_dir, remote_store.logs_subdir)
             logger = TensorBoardLogger(logs_path)
 
-            ckpt_path = os.path.join(run_output_dir, remote_store.checkpoint_filename)
-            os.makedirs(ckpt_path, exist_ok=True)
+            # TODO: find out a way to use ckpt_path created from remote store, but all other parameters ingest from estimator config
+            # ckpt_path = os.path.join(run_output_dir, remote_store.checkpoint_filename)
+            # os.makedirs(ckpt_path, exist_ok=True)
+            # model_checkpoint_callback = ModelCheckpoint(dirpath=ckpt_path)
+            # callbacks.append(model_checkpoint_callback)
 
-            # disable checkpoint call back for now, waiting for the fix of
-            # https://github.com/PyTorchLightning/pytorch-lightning/issues/6343
-            checkpoint_callback = None# ModelCheckpoint(dirpath=ckpt_path)
+            is_model_checkpoint_callback_exist = False
+            if callbacks is not None:
+                for cb in callbacks:
+                    if isinstance(cb, ModelCheckpoint):
+                        is_model_checkpoint_callback_exist = True
+                        break
 
             model = deserialize(serialized_model)
+
+            cuda_available = torch.cuda.is_available()
+            if cuda_available:
+                # Horovod: pin GPU to local rank or the assigned GPU from spark.
+                torch.cuda.set_device(_get_assigned_gpu_or_default(default=hvd.local_rank()))
+                # Move model to GPU.
+                model.cuda()
+
+            if num_gpus is None:
+                num_gpus = 1 if cuda_available else 0
+
             kwargs = {'accelerator': 'horovod',
-                'gpus': (1 if torch.cuda.is_available() else 0),
-                'callbacks': callbacks,
-                'max_epochs': epochs,
-                'limit_train_batches': train_percent,
-                'limit_val_batches': val_percent,
-                'logger': logger,
-                'checkpoint_callback': checkpoint_callback,
-                'resume_from_checkpoint': (last_ckpt_file if ckpt_bytes else None),
-                'num_sanity_val_steps': 0
-            }
+                      'gpus': num_gpus,
```

Comment:
It looks like lightning support num_gpus per process? How is gradients allreduce working in that case?

Reply:
It is defined [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L201)
And used to get device id [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/a584196abf820179adb0758ef67ddae91c44e7bc/pytorch_lightning/utilities/device_parser.py#L46)

Reply:
This `num_gpus` means num_gpus per node.

Seeing this https://github.com/PyTorchLightning/pytorch-lightning/blob/a584196abf820179adb0758ef67ddae91c44e7bc/pytorch_lightning/trainer/connectors/accelerator_connector.py#L625

Horovod backend doesn't support `num_gpus` per node?

Reply:
And this: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/accelerator_connector.py#L599

It is saying no need to set n_gpus for horovod